### ISSUE TITLE
Quotes sometimes needs a name in Scala 3 macros

### ISSUE THREAD

**Issue by markehammons** [lampepfl/dotty#13905]
## Compiler version

3.1.0

## Minimized code
note: I'll add minimized code later, this bug seems hard to trigger

![image](https://user-images.githubusercontent.com/2609471/140666186-b0087c7e-5905-46d1-83ea-74f2d3f17575.png)
as you see, there's an error highlighted in this code.

Giving `Quotes` a name resolves the error.

![image](https://user-images.githubusercontent.com/2609471/140666237-8309457b-8fc1-4765-bbc5-7961c1619083.png)

## Output

```scala
core/src/io/gitlab/mhammons/slinc/nativepieces.scala:362:16
[E]      Found:    (x$1 : (String, quoted.Type[? <: AnyKind]))
[E]      Required: quoted.Quotes
[E]      L362:                '{
```

## Expectation

I shouldn't need a name for Quotes here.

**Comment by markehammons** [comment]
Sorry for the wait on this, been busy as a bee on my own projects. So far, the common trigger for this bug seems to be using quotes inside of a partial function.

**Comment by markehammons** [comment]
https://github.com/markehammons/macro-bug-13905

Here's a minimized version of the bug. The issue appears when doing a `map-match`.

**Comment by bishabosha** [comment]
^ error with the anonymous using parameter:
```scala
-- Error:
5 |        '{ ${ Expr(Type.show[r]) } }
  |        ^
  |Found:    quoted.Type[?1.CAP]
  |Required: quoted.Quotes
  |
  |where:    ?1 is an unknown value of type scala.runtime.TypeBox[Nothing, AnyKind]
```

**Comment by smarter** [comment]
Fixed in 3.3.1-RC1 and up as reported in https://github.com/lampepfl/dotty/issues/18059
